optimize/cmaes.go

```diff
@@ -0,0 +1,382 @@
+// InitMean) to have a better chance of finding the global minimum.
+//
+// The CMA-ES-Chol algorithm differs from the standard CMA-ES algorithm in that
+// it directly updates the Cholesky decomposition of the normal distribution.
+// This changes the runtime is O(dim^2*pop) rather than O(dim^3). The evolution
+// of the multi-variate normal will be similar to the baseline CMA-ES algorithm,
+// but the covariance update equation is not identical.
+//
+// For more information about the CMA-ES algorithm, see
+//  https://en.wikipedia.org/wiki/CMA-ES
+//  https://arxiv.org/pdf/1604.00772.pdf
+type CmaEsChol struct {
+	// InitStepSize sets the initial size of the covariance matrix adaptation.
+	// If InitStepSize is 0, a default value of 0.5 is used. InitStepSize cannot
+	// be negative, or CmaEsChol will panic.
+	InitStepSize float64
+	// Population sets the population size for the algorithm. If Population is
+	// 0, a default value of 4 + math.Floor(3*math.Log(float64(dim))) is used.
+	// Population cannot be negative or CmaEsChol will panic.
+	Population int
+	// InitMean is the initial mean of the multivariate normal for sampling
+	// input locations. If InitMean is nil, the zero vector is used. If InitMean
+	// is not nil, it must have length equal to the problem dimension.
+	InitMean []float64
+	// InitCholesky specifies the Cholesky decomposition of the covariance
+	// matrix for the initial sampling distribution. If InitCholesky is nil,
+	// a default value of I is used. If it is non-nil, then it must have
+	// InitCholesky.Size() be equal to the problem dimension.
+	InitCholesky *mat.Cholesky
+	// StopLogDet sets the threshold for stopping the optimization if the
+	// distribution becomes too peaked. The log determinant is a measure of the
+	// (log) "volume" of the normal distribution, and when it is too small
+	// the samples are almost the same. If the log determinant of the covariance
+	// matrix becomes less than StopLogDet, the optimization run is concluded.
+	// If StopDeterminant is 0, a default value of dim*log(1e-16) is used.
+	// If StopDeterminant is NaN, the stopping criteria is not used, though
+	// this can cause numeric instabilities in the algorithm.
+	StopLogDet float64
+	// ForgetBest, when true, does not track the best overall function value found,
+	// instead returning the new best sample in each iteration. If ForgetBest
+	// is false, then the minimum value returned will be the lowest across all
+	// iterations, regardless of when that sample was generated.
+	ForgetBest bool
+	// Src allows a random number generator to be supplied for generating samples.
+	// If Src is nil the generator in math/rand is used.
+	Src *rand.Rand
+
+	// Fixed algorithm parameters.
+	dim                 int
+	pop                 int
+	weights             []float64
+	muEff               float64
+	cc, cs, c1, cmu, ds float64
+	eChi                float64
+
+	// Function data.
+	xs *mat.Dense
```

Comment:
I disagree. Each row has exactly the same number of columns, which is not guaranteed by `[][]float64` or a returning method (it would be better as a multi-dim slice but Go doesn't have those). It's very common to have a list of samples stored as a matrix, and I think that fits here.

Reply:
>an exported method

I meant 'unexported'. Sorry.

I know it is common to use Dense as a 2D array storage without needing the matrix semantics but I still disagree with it (unless it is justified by usage in an API that takes a matrix, for example stat.PC, but that is not the case here). A downside of Dense is that one has to specify which way are the samples and which way are the variables. For this `[][]float64` (or a dedicated type or something else) would be much clearer and since it is our internal code we can make sure that all the slices would have the same length. But if you wish to keep it as Dense, then it's fine by me.

Reply:
I think it is better with Dense.

> A downside of Dense is that one has to specify which way are the samples and which way are the variables.

It is a downside, unless we're consistent in how we use it, as in "all Gonum functions have the samples as rows", which we do so far (PCA, Covariance, etc.). This also has the advantage of being the same indexing as `[][]float64`, but with the additional safety and storage guarantees.